libp2p/protocols/mix/mix_protocol.nim

```diff
@@ -72,21 +73,46 @@ proc cryptoRandomInt(rng: ref HmacDrbgContext, max: int): Result[int, string] =
   let res = rng[].generate(uint64) mod uint64(max)
   ok(res.int)
 
-proc handleMixNodeConnection(
-    mixProto: MixProtocol, conn: Connection
-) {.async: (raises: [LPStreamError, CancelledError]).} =
-  when defined(enable_mix_benchmarks):
-    let metadataBytes = newSeqUninit[byte](MetadataSize)
-    await conn.readExactly(addr metadataBytes[0], MetadataSize)
-
-  let receivedBytes =
-    try:
-      await conn.readLp(PacketSize)
-    except CancelledError as exc:
-      raise exc
-    finally:
-      await conn.close()
+proc getConn(
+    mixProto: MixProtocol,
+    pid: PeerId,
+    addrs: seq[MultiAddress],
+    codecs: seq[string],
+    forceNewStream: bool = false,
+): Future[Connection] {.async: (raises: [DialFailedError, CancelledError]).} =
+  if mixProto.sendPool.hasKey(pid):
+    if forceNewStream:
+      mixProto.sendPool.del(pid)
+    else:
+      try:
+        return mixProto.sendPool[pid]
+      except KeyError:
+        raiseAssert "checked with hasKey"
+  let c = await mixProto.switch.dial(pid, addrs, codecs)
+  mixProto.sendPool[pid] = c
+  return c
+
+proc writeLp(
+    mixProto: MixProtocol,
+    pid: PeerId,
+    addrs: seq[MultiAddress],
+    codecs: seq[string],
+    payload: seq[byte],
+) {.async: (raises: [DialFailedError, LPStreamError, CancelledError]).} =
+  let c = await mixProto.getConn(pid, addrs, codecs)
+  try:
+    await c.writeLp(payload)
+  except LPStreamClosedError, LPStreamResetError, LPStreamRemoteClosedError,
+      LPStreamConnDownError:
+    let c = await mixProto.getConn(pid, addrs, codecs, forceNewStream = true)
+    await c.writeLp(payload)
```

Comment:
Is there not a check we can do in `c` to avoid calling `getConn` and `c.writeLp` twice? e.g.

```nim
var c = await mixProto.getConn(pid, addrs, codecs)
if c.isClosed() or c.isReset() or c.isRemoteClosed():
  await mixProto.getConn(pid, addrs, codecs, forceNewStream = true)
await c.writeLp(payload)
```

Reply:
makes sense